KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,463 @@
+    type == "adc" && return s.ADC[1]
+    return Sequence()
+end
+
+"""
+    type_blk_obj_id = get_type_blk_obj_id(seq::Sequence, type::String, unique_obj_id::Vector)
+
+Returns the vector `type_blk_obj_id` of vectors [blk, obj, id] for all the blocks `blk` of a
+sequence `seq` according to the `type` in ["rf", "gx", "gy", "gz", "adc"] chosen. It is
+neccessary to add the input vector `unique_obj_id` which contains the uniques objects (RF,
+Grad, or ADC) with its repective ID.
+"""
+function get_type_blk_obj_id(seq::Sequence, type::String, unique_obj_id::Vector)
+    type_blk_obj_id = [[blk, get_obj(s, type), 0] for (blk, s) in enumerate(seq)]
+    for boi in type_blk_obj_id
+        for (obj_unique, id_unique) in unique_obj_id
+            if boi[2] ≈ obj_unique
+                boi[3] = id_unique
+            end
+        end
+    end
+    return type_blk_obj_id
+end
+
+"""
+    output = magsign(v)
+
+Returns the extreme value (positive or negative) which has the maximum absolute value.
+"""
+function magsign(v)
+    a, b = extrema(v)
+    return -a > b ? a : b
+end
+
+"""
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, id_shape_cnt = get_rfunique(rfunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+
+Returns the unique shapes for the magnitude, angle and time of the "rfunique_obj_id" vector.
+Requires an initial integer counter "id_shape_cnt" to asign IDs incrementally.
+"""
+function get_rfunique(rfunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+    # Find the unique shapes (magnitude, phase and time shapes) and assign IDs
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id = [], [], []
+    for (obj, _) in rfunique_obj_id
+        shape_abs = abs.(obj.A) / maximum(abs.(obj.A))
+        if all([
+            !(
+                length(shape_abs) == length(shape_abs_unique) &&
+                shape_abs ≈ shape_abs_unique
+            ) for (shape_abs_unique, _) in rfunique_abs_id
+        ])
+            push!(rfunique_abs_id, [shape_abs, id_shape_cnt])
+            id_shape_cnt += 1
+        end
+        shape_ang = mod.(angle.(obj.A), 2π) / 2π
+        if all([
+            !(
+                length(shape_ang) == length(shape_ang_unique) &&
+                all(shape_ang - shape_ang_unique .≈ shape_ang[1] - shape_ang_unique[1])
+            ) for (shape_ang_unique, _) in rfunique_ang_id
+        ])
+            push!(rfunique_ang_id, [shape_ang, id_shape_cnt])
+            id_shape_cnt += 1
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T]) / seq.DEF["RadiofrequencyRasterTime"]
+            if all([
+                !(
+                    length(shape_tim) == length(shape_tim_unique) &&
+                    shape_tim ≈ shape_tim_unique
+                ) for (shape_tim_unique, _) in rfunique_tim_id
+            ])
+                push!(rfunique_tim_id, [shape_tim, id_shape_cnt])
+                id_shape_cnt += 1
+            end
+        end
+    end
```

Comment:
It looks like it could be greatly simplified. Be careful with the phase comparisons, I suggest to do

$$
\left| \frac{1}{N}\sum_{n}^{N}\exp(-i\phi_1(n)) \cdot \exp(i\phi_2(n)) \right|  == 1,
$$

with $\phi_1$ the phase waveform of `rf1` and $\phi_2$ the phase waveform of the `rf2`.

Reply:
Now is simpler. Check here bcf5515